### DISCUSSION TITLE
HTML parameters are not hydrated when the server and browser version of a variable differ.

### DISCUSSION THREAD

**Discussion by mildfuzz** [vercel/next.js#28084] [Help] [answered]
### What version of Next.js are you using?

11.1.0

### What version of Node.js are you using?

14.1.0

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

npm run dev

### Describe the Bug

Given this trivial page:

```typescript
import type { NextPage } from 'next'
const isBrowser = () => !!global?.window
const data = isBrowser() ? 'browser' : 'server'
const Home: NextPage = () => {  return (
      <p data-ref={data}>{data}</p>
  )
}
export default Home
```

I see this HTML
```html
<p data-ref="server">browser</p>
```
It seems that when the browser hydrates the HTML, it doesn't touch the parameters

### Expected Behavior

I believe we should see the following HTML:

```html
<p data-ref="browser">browser</p>
```

### To Reproduce

Run the following as page/index.tsx

```typescript
import type { NextPage } from 'next'
const isBrowser = () => !!global?.window
const data = isBrowser() ? 'browser' : 'server'
const Home: NextPage = () => {  return (
      <p data-ref={data}>{data}</p>
  )
}
export default Home
```

**Comment by timneutkens** [comment, marked as answer]
This is expected as during hydration the data has to match up with the server-rendered HTML. The correct way to let React know that there's a change to the data is this:

```
import type { NextPage } from 'next'
import {useEffect, useState} from 'react'

const Home: NextPage = () => {  
  const [isBrowser, setBrowser] = useState(false)
  useEffect(() => setBrowser(true), [])
  return (
      <p data-ref={isBrowser ? 'browser' : 'server}>{data}</p>
  )
}
export default Home
```

**Reply by mildfuzz** [reply]
Thanks for quick reply. 

The real world issue I have is that, when using recoil-persist, the default recoil state is different on server and browser, causing inconsistencies which led me to identify this.

**Reply by mildfuzz** [reply]
From what I can gather, the real root cause is Atom Effect

**Reply by mildfuzz** [reply]
What I am curious about, re: next.js, is if it needs to match the server HTML, is it aware when it does not? could it trigger a re-render itself?

**Reply by mildfuzz** [reply]
@timneutkens are you able to expand for my understanding?

**Reply by timneutkens** [reply]
Next.js uses the particular React APIs for SSR/hydration, it does not match the DOM by itself:
https://reactjs.org/docs/react-dom.html#hydrate
https://reactjs.org/docs/react-dom-server.html#rendertostring

During hydration you should be able to see warnings in the browser console when it does not match up coming from React, that might not be the case for `data-` though, not sure about that particular case.

**Reply by mildfuzz** [reply]
interesting, so potentially could get a browser warning, but not an event or something we can respond to
